storage/server: Add reserve RPC

BramGruneir opened the pull request:
This is the receiving side of the reservation system.

6 commits in this PR:
- Add reserved info to store stats
- Add reserve RPC
- Update Reserve() to limit reservations.
- Hook up bookie.Fill to ApplySnapshot.
- Rename `Approved` to `Reserved` in `ReservationResponse`
- Misc style fixes, rename `booking` to `reservation`

The next step will be to hook up the allocator/store pool to make the rpc call before trying to rebalance.  

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6904)

<!-- Reviewable:end -->

BramGruneir committed:
storage: Add reservation info to the store

Gossiped store descriptors will now include reservation information baked in.
Added new store stats to expose reservation stats.

BramGruneir commented:
@mrtracy can you take a quick look too?  Just to double check the new metrics.

cuongdo commented:
meta-comment: I think that this PR should be bundled with the reservation RPC (maybe as two separate commits in the same PR). It's hard for me to completely see how things fit together without seeing the consumer of the reservation data. Also, it seems likely that while implementing the RPC, you might have to tweak `StoreDescriptor` again.

<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> @mrtracy can you take a quick look too?  Just to double check the new metrics.
</details>


---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 449 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIdO6adeSU94ZLVadM4:-KIdO6aeZ1FaZW_yxpVu:894873954) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L449)):*
> ```Go
> 	capacity        *metric.Gauge
> 	available       *metric.Gauge // Does not include reserved bytes.
> 	reserved        *metric.Counter
> ```

following the naming convention of the other members of this struct, this should be `reservedBytes`

---

*[storage/store.go, line 1636 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIdOWl_EXFaTIp10k8S:-KIdOWla31flpZbD0DJ_:-1254021789) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L1636)):*
> ```Go
> // DescriptorWithoutReservations returns a StoreDescriptor including current
> // store capacity information without including reservations.
> func (s *Store) DescriptorWithoutReservations() (*roachpb.StoreDescriptor, error) {
> ```

I'm nervous about having the semantics of the `StoreDescriptor` fields depend on which method was used to retrieve it. I'd prefer that the `StoreDescriptor` have separate fields for reserved replicas & bytes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-:-KIdNNxtBRd_XTwaF4cF:1817562022)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
*[storage/store.go, line 449 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIdO6adeSU94ZLVadM4:-KIdP71N00xBGLKVPPzH:1442035844) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L449)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> following the naming convention of the other members of this struct, this should be `reservedBytes`
</details>
actually, never mind... you're following the convention set by `available`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
I see your point.  I can get the rpc in as the next commit before merging.

<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> meta-comment: I think that this PR should be bundled with the reservation RPC (maybe as two separate commits in the same PR). It's hard for me to completely see how things fit together without seeing the consumer of the reservation data. Also, it seems likely that while implementing the RPC, you might have to tweak `StoreDescriptor` again.
</details>


---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 1636 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIdOWl_EXFaTIp10k8S:-KIdbTlOaok4J8BB4KK7:-286672423) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L1636)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> I'm nervous about having the semantics of the `StoreDescriptor` fields depend on which method was used to retrieve it. I'd prefer that the `StoreDescriptor` have separate fields for reserved replicas & bytes.
</details>
I originally had that approach, but capacity is used in engine, and the engine itself doesn't know about reservations (and shouldn't).  To get the number of reservations, we can use the store metrics.  I think we're going to want to gossip store metrics anyway, since they contain a lot more data.  But that will be an optimization after the reservation system is in place.

For now, all the other stores only need to know is that there is less capacity available and it has more ranges.  Regardless of if they are reserved or real.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-:-KIdcMzGpuogdWhqASr5:1671921279)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 448 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIe8W5NfSLOur2S0Yt1:-KIe8W5NfSLOur2S0Yt2:-146092574) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L448)):*
> ```Go
> 	lastUpdateNanos *metric.Gauge
> 	capacity        *metric.Gauge
> 	available       *metric.Gauge // Does not include reserved bytes.
> ```

This comment is ambiguous: `available` doesn't include used bytes, but I don't think that's the sense in which you're using "include" in this comment.

---

*[storage/store.go, line 505 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIe9NJyh0qbNreIcQ8c:-KIe9NJyh0qbNreIcQ8d:-1825408541) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L505)):*
> ```Go
> 		capacity:                 storeRegistry.Gauge("capacity"),
> 		available:                storeRegistry.Gauge("capacity.available"),
> 		reserved:                 storeRegistry.Counter("capacity.reserved"),
> ```

Why is this a counter and not a gauge (like `available` is)? I think both reserved stats need to be gauges and not counters, because they are not only incremented - reservations are lost on restart and they reset to zero instead of continuing with their previous value.

---

*[storage/store.go, line 1636 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIdOWl_EXFaTIp10k8S:-KIeA3X89uzkAUxRZPTP:1433105348) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L1636)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I originally had that approach, but capacity is used in engine, and the engine itself doesn't know about reservations (and shouldn't).  To get the number of reservations, we can use the store metrics.  I think we're going to want to gossip store metrics anyway, since they contain a lot more data.  But that will be an optimization after the reservation system is in place.
> 
> For now, all the other stores only need to know is that there is less capacity available and it has more ranges.  Regardless of if they are reserved or real.
</details>
I agree with @cuongdo; I'd much rather introduce new fields to the store descriptor for reserved ranges and bytes, and have the bookie and rebalancing queue be the only things that add reservations to current usage.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

BramGruneir committed:
server: Adds the reserve rpc

The reserve rpc forwards reserve calls via the node to the correct store
requesting reservation.

BramGruneir committed:
storage: add max reservations and size limits for reserve calls

This change limits the reservations available per store. It adds a constraints
on the total number of reserved bytes and replicas.
It also checks to make sure there is enough free space on the hard drive to
accommodate the reservation.

BramGruneir commented:
Still a WIP.

BramGruneir committed:
storage: fill reservations when a snapshot arrives

BramGruneir commented:
Review status: 0 of 9 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 448 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIe8W5NfSLOur2S0Yt1:-KJbSvKKYQRlelLCyyvH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L448)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This comment is ambiguous: `available` doesn't include used bytes, but I don't think that's the sense in which you're using "include" in this comment.
</details>
Done.

---

*[storage/store.go, line 505 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIe9NJyh0qbNreIcQ8c:-KJHff2gi3a2YV3bT7Ev:-2146970724) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L505)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why is this a counter and not a gauge (like `available` is)? I think both reserved stats need to be gauges and not counters, because they are not only incremented - reservations are lost on restart and they reset to zero instead of continuing with their previous value.
</details>
I've updated the code to deal with restarting.  @mrtracy specifically mentioned that we should be using counters and not gauges.

---

*[storage/store.go, line 1636 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIdOWl_EXFaTIp10k8S:-KJbgc7Z32loSiXTuydD:-2029414338) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L1636)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I agree with @cuongdo; I'd much rather introduce new fields to the store descriptor for reserved ranges and bytes, and have the bookie and rebalancing queue be the only things that add reservations to current usage.
</details>
I've removed this for now.  I think the best course of action will be to either gossip the metrics themselves (they include a store descriptor already), or to add the metrics to the store descriptor.  @mrtracy and I are reviewing the options.  

I think some kind of whitelisted subset of the metrics (those that are needed for rebalancing) would be all we need.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Ok, I think this is good to go!
PTAL!

cuongdo commented:
Reviewed 5 of 7 files at r3, 3 of 3 files at r4.
Review status: 8 of 9 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 831 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfnxuyD35q9DjA9QtB:-KJfnxuyD35q9DjA9QtC:-929319700) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/roachpb/api.proto#L831)):*
> ```protobuf
>   optional int64 range_id = 4 [(gogoproto.nullable) = false,
>       (gogoproto.customname) = "RangeID", (gogoproto.casttype) = "RangeID"];
>   optional int64 range_size = 5 [(gogoproto.nullable) = false];
> ```

Do we really need this? Can we be conservative and round up to `range_max_bytes` for the zone? If the receiver of the reservation request can't handle the range once it's filled, should it even approve the reservation?

---

*[roachpb/api.proto, line 836 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfoiQI0jMAMPTUsu-A:-KJfoiQI0jMAMPTUsu-B:-1808228787) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/roachpb/api.proto#L836)):*
> ```protobuf
> // A ReservationResponse is the repoonse returned from a ReservationRequest.
> message ReservationResponse {
>   optional bool approved = 1 [(gogoproto.nullable) = false];
> ```

nit: maybe change this to `reserved`? `approved` sounds slightly weaker than `reserved`

---

*[storage/replica_raftstorage.go, line 715 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfpxQRomSnp7RVQekf:-KJfpxQRomSnp7RVQekg:287484355) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_raftstorage.go#L715)):*
> ```Go
> 
> 		// Fill the reservation if there was any one for this range.
> 		_ = r.store.bookie.Fill(rangeID)
> ```

Are there any cases where this line of code will be executed for anything other than `ChangeReplica`? For example, if a node crashes and restarts later, are there cases where it will apply a snapshot to catch up to the current Raft index?

---

*[storage/replica_test.go, line 5690 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJftN3dENslcWclgNwr:-KJftN3dENslcWclgNws:-1425426438) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_test.go#L5690)):*
> ```Go
> 		defer tc.store.bookie.mu.Unlock()
> 		if e, a := expected, len(tc.store.bookie.mu.resByRangeID); e != a {
> 			t.Fatalf("reservation still exist - expected:%d, actual:%d", e, a)
> ```

s/exist/exists/

---

*[storage/replica_test.go, line 5701 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfqgntNffeYH_Fhs78:-KJfqgntNffeYH_Fhs79:360798553) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_test.go#L5701)):*
> ```Go
> 	}
> 
> 	tc.store.metrics.available.Update(1024 ^ 3)
> ```

what is the significance of`1024 ^ 3`? edit: I see now, though it seems like you should use a constant here

---

*[storage/reservation.go, line 141 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJftszJDgX5YwVtFwKr:-KJftszJDgX5YwVtFwKs:-456554666) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/reservation.go#L141)):*
> ```Go
> 	// Store `available` in case it changes between if and log.
> 	available := b.metrics.available.Value()
> 	if b.mu.size+(req.RangeSize*2) > available {
> ```

Do we need the `*2` here if you always reserve `range_max_bytes` and never choose a replication target that has less than 5% free space (the latter is done now)?

---

*[storage/reservation_test.go, line 64 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfrZVHo5ViusSv-lRJ:-KJfrZVHo5ViusSv-lRK:-133132754) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/reservation_test.go#L64)):*
> ```Go
> func TestReserve(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> 	stopper, _, b := createTestBookie(time.Hour, 20, 20*1024^3)
> ```

what is the significance of `20` and `20*1024^3`?

---

*[storage/store.go, line 505 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIe9NJyh0qbNreIcQ8c:-KJflGKWz2AkWftNu6XM:2034106305) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L505)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I've updated the code to deal with restarting.  @mrtracy specifically mentioned that we should be using counters and not gauges.
</details>
Why is this? Is it so that we can see rates of reservations in the UI?

---

*[storage/store.go, line 76 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfsOjq0egTFgN5qnbS:-KJfsOjq0egTFgN5qnbT:-1013107350) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/store.go#L76)):*
> ```Go
> 	// them reactive to the store in some way.
> 	maxReservations  = 100
> 	maxReservedBytes = 1024 ^ 3
> ```

isn't `^` a bitwise xor? did you mean `math.Pow(1024, 3)` (probably doesn't work outside a function) or `1024 * 1024 * 1024`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

BramGruneir committed:
server/storage: change reservation response from `approved` to `reserved`

BramGruneir commented:
Review status: 2 of 9 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 831 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfnxuyD35q9DjA9QtB:-KJfwYrCWQ18VV1vGDY4:1357042286) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/roachpb/api.proto#L831)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Do we really need this? Can we be conservative and round up to `range_max_bytes` for the zone? If the receiver of the reservation request can't handle the range once it's filled, should it even approve the reservation?
</details>
We do.  Ranges can be larger than the expected size and frequently are much smaller.  We want to allow a larger number of incoming snapshots if they're all quite small.

---

*[roachpb/api.proto, line 836 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfoiQI0jMAMPTUsu-A:-KJg-z9LSF6EMcQBk3bX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/roachpb/api.proto#L836)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> nit: maybe change this to `reserved`? `approved` sounds slightly weaker than `reserved`
</details>
Done.

---

*[storage/replica_raftstorage.go, line 715 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfpxQRomSnp7RVQekf:-KJfwrz7xbdBHRKL7q57:668281552) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_raftstorage.go#L715)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Are there any cases where this line of code will be executed for anything other than `ChangeReplica`? For example, if a node crashes and restarts later, are there cases where it will apply a snapshot to catch up to the current Raft index?
</details>
Yes, it will be called when applying a catch up snapshot, but since no reservation will exist, it will have no effect.

---

*[storage/replica_test.go, line 5690 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJftN3dENslcWclgNwr:-KJg05_FAnumfLfVj0hz:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_test.go#L5690)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> s/exist/exists/
</details>
Done.

---

*[storage/replica_test.go, line 5701 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfqgntNffeYH_Fhs78:-KJg-G-QLAWJMj95hLji:1706342937) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_test.go#L5701)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> what is the significance of`1024 ^ 3`? edit: I see now, though it seems like you should use a constant here
</details>
Done. 

---

*[storage/reservation.go, line 141 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJftszJDgX5YwVtFwKr:-KJg-KOczyRdkRhhFbuy:530178293) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/reservation.go#L141)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Do we need the `*2` here if you always reserve `range_max_bytes` and never choose a replication target that has less than 5% free space (the latter is done now)?
</details>
Is it needed?  I'm not sure.  It doesn't hurt to have it.  Especially for larger ranges.  And since we can't fully rely on range_max_bytes.

---

*[storage/reservation_test.go, line 64 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfrZVHo5ViusSv-lRJ:-KJg-feMoJfX_OiWGnS-:-1087359574) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/reservation_test.go#L64)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> what is the significance of `20` and `20*1024^3`?
</details>
changed it to a constant.

---

*[storage/store.go, line 505 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIe9NJyh0qbNreIcQ8c:-KJfx6-ZUxIOqSyOWKRb:-1808232678) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L505)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Why is this? Is it so that we can see rates of reservations in the UI?
</details>
In both the UI for debugging but also I would like to gossip this value so we can make smarter decisions.

---

*[storage/store.go, line 76 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfsOjq0egTFgN5qnbS:-KJg-mAhCE6vU5hiU3ZT:1453884825) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/store.go#L76)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> isn't `^` a bitwise xor? did you mean `math.Pow(1024, 3)` (probably doesn't work outside a function) or `1024 * 1024 * 1024`?
</details>
it is.  I was wondering why my numbers were a bit strange during testing.  Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 4 of 7 files at r3, 3 of 3 files at r4, 1 of 3 files at r5, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[roachpb/api.proto, line 831 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfnxuyD35q9DjA9QtB:-KJg-VikEr0MqAV_VLkP:-867631) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/roachpb/api.proto#L831)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> We do.  Ranges can be larger than the expected size and frequently are much smaller.  We want to allow a larger number of incoming snapshots if they're all quite small.
</details>
It may makes sense to use the larger of the the maximum size and the current size (this may unnecessarily serialize small ranges, but that doesn't seem like a big deal), but either way this will need to be sent as a part of the request - there's not enough information here for the recipient to be able to look up the right zone.

---

*[storage/replica_raftstorage.go, line 715 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfpxQRomSnp7RVQekf:-KJg09dd9nzzkHiv6ddZ:-1751344096) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_raftstorage.go#L715)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Are there any cases where this line of code will be executed for anything other than `ChangeReplica`? For example, if a node crashes and restarts later, are there cases where it will apply a snapshot to catch up to the current Raft index?
</details>
Yes, that's possible. But as long as `Fill` is a noop when there is no actual reservation it should be fine. It's difficult to tell the difference otherwise.

Also, I wouldn't use `_ =` when there are no other return values unless you're silencing `errcheck`

---

*[storage/replica_test.go, line 5690 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJftN3dENslcWclgNwr:-KJgFflr-HSbi-s0MZvs:-2044385253) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_test.go#L5690)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> s/exist/exists/
</details>
The message "still exists" only makes sense if `expected` is 0. 

---

*[storage/replica_test.go, line 5720 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgGHXew5xOCZq1ewl7:-KJgGHXew5xOCZq1ewl8:768388636) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_test.go#L5720)):*
> ```Go
> 	}
> 	checkReservations(t, 1)
> 	b := tc.engine.NewBatch()
> ```

`defer b.Close()`

---

*[storage/reservation.go, line 96 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgKKdSA7m5CSI00n6M:-KJgKKdTKjEwUfK5N42X:601674832) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/reservation.go#L96)):*
> ```Go
> 	}
> 	b.mu.resByRangeID = make(map[roachpb.RangeID]*booking)
> 	b.metrics.reserved.Clear()
> ```

How does this help? Store metrics are not loaded from persistent storage so there's nothing to clear at this point.

---

*[storage/reservation.go, line 136 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgKiLLIl1lbRJvmOXe:-KJgKiLLIl1lbRJvmOXf:1364684411) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/reservation.go#L136)):*
> ```Go
> 	}
> 
> 	// Can we accomidate the requested number of bytes (doubled for safety) on
> ```

s/accomidate/accomodate/

---

*[storage/store.go, line 505 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIe9NJyh0qbNreIcQ8c:-KJgYobNfDckCmtISxEq:-1994131233) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L505)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> In both the UI for debugging but also I would like to gossip this value so we can make smarter decisions.
</details>
@mrtracy, why do you think this should be a counter? It seems like a natural gauge to me. (same for `reservedReplicaCount` and the existing `replicaCount`)

---

*[storage/store.go, line 76 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfsOjq0egTFgN5qnbS:-KJgJDzQnCFnRlNuH-kd:-595218617) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/store.go#L76)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> it is.  I was wondering why my numbers were a bit strange during testing.  Fixed.
</details>
Or `1<<30`, which appears to be the more common style in our code. (s/gigibyte/GiB/)

I would make these numbers really small to start to prioritize stability (say 5 and 250MB); we can consider increasing them if rebalancing is still too slow (or when we have streaming snapshots).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 7 files at r3, 1 of 3 files at r5, 4 of 5 files at r6.
Review status: all files reviewed at latest revision, 25 unresolved discussions, all commit checks successful.

---

*[storage/replica_test.go, line 5690 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJftN3dENslcWclgNwr:-KJgcteHRaButs5dRyw3:5172523) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_test.go#L5690)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The message "still exists" only makes sense if `expected` is 0. 
</details>
+1

---

*[storage/replica_test.go, line 5698 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJggY4P-enoMtz5h6ti:-KJggY4P-enoMtz5h6tj:-1366659602) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/replica_test.go#L5698)):*
> ```Go
> 	snap, err := firstRng.GetSnapshot()
> 	if err != nil {
> 		t.Fatalf("can't get snapshot: %s", err)
> ```

just wondering, why not `t.Fatal(err)`? there's already a line number

---

*[storage/replica_test.go, line 5722 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJggiDTE_o22kz4oaJw:-KJggiDTE_o22kz4oaJx:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/replica_test.go#L5722)):*
> ```Go
> 	b := tc.engine.NewBatch()
> 	if _, err := firstRng.applySnapshot(b, snap); err != nil {
> 		t.Fatalf("couldn't apply the snapshot: %s", err)
> ```

ditto

---

*[storage/replica_test.go, line 5725 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJggiy7JymQ30JBWL1N:-KJggiy7JymQ30JBWL1O:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/replica_test.go#L5725)):*
> ```Go
> 	}
> 	if err := b.Commit(); err != nil {
> 		t.Fatalf("couldn't commit the batch: %s", err)
> ```

ditto

---

*[storage/reservation.go, line 96 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgKKdSA7m5CSI00n6M:-KJggzobbQ6LJTbIDnLw:-1027311471) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/reservation.go#L96)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> How does this help? Store metrics are not loaded from persistent storage so there's nothing to clear at this point.
</details>
this also mutates the argument; is that intended?

---

*[storage/reservation.go, line 87 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJggvjIhSpcBRh05lme:-KJggvjIhSpcBRh05lmf:-1131678282) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation.go#L87)):*
> ```Go
> 	maxReservedBytes int64,
> 	stopper *stop.Stopper,
> 	metrics *storeMetrics) *bookie {
> ```

doesn't conform to the style guide

---

*[storage/reservation.go, line 95 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJghQ2aGZRXy72zum4s:-KJghQ2aGZRXy72zum4t:601122650) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation.go#L95)):*
> ```Go
> 		metrics:            metrics,
> 	}
> 	b.mu.resByRangeID = make(map[roachpb.RangeID]*booking)
> ```

is it a reservation or a booking? this naming is confusing. Also, why pointers in this map?

---

*[storage/reservation.go, line 102 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgh55ueVq0fPQg-44f:-KJgh55ueVq0fPQg-44g:1963558233) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation.go#L102)):*
> ```Go
> }
> 
> // Reserve a new replica. Reservations can be rejected due to having too many
> ```

this comment should explain what a rejected reservation looks like to the caller.

---

*[storage/reservation.go, line 107 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJghHC8ZCz-qSiWbUEy:-KJghHC8ZCz-qSiWbUEz:602634864) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation.go#L107)):*
> ```Go
> 	b.mu.Lock()
> 	defer b.mu.Unlock()
> 	if olderReservation, ok := b.mu.resByRangeID[req.RangeID]; ok {
> ```

is this a 

---

*[storage/reservation.go, line 114 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJghtP0E0qTDwpexasJ:-KJghtP0E0qTDwpexasK:-587093832) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation.go#L114)):*
> ```Go
> 			// new one.
> 			if log.V(2) {
> 				log.Infof("updating existing reservation for rangeID:%d, %+v", req.RangeID,
> ```

throughout: can you give the booking/reservation and the reservation request useful string methods rather than always logging them `%+v`? Unless every field in both is typically useful, in which case ignore this comment.

---

*[storage/reservation.go, line 130 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJghcciWw5bAyQgWdFB:-KJghcciWw5bAyQgWdFC:1976898556) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation.go#L130)):*
> ```Go
> 	if len(b.mu.resByRangeID) > b.maxReservations {
> 		if log.V(2) {
> 			log.Infof("could not book reservation %+v, too many reservations already %d",
> ```

did you mean to log the number of live reservations here?

---

*[storage/reservation_test.go, line 33 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgi93Up-VcTDXVwO_K:-KJgi93Up-VcTDXVwO_L:-19230703) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation_test.go#L33)):*
> ```Go
> 	reservationTimeout time.Duration,
> 	maxReservations int,
> 	maxReservedBytes int64) (*stop.Stopper, *hlc.ManualClock, *bookie) {
> ```

does not conform to style guide

---

*[storage/reservation_test.go, line 172 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgiI2fLaeViYqirojR:-KJgiI2fLaeViYqirojS:794413981) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation_test.go#L172)):*
> ```Go
> 
> 	if b.Reserve(overfilledReq).Reserved {
> 		t.Errorf("overfilled the reservations system")
> ```

throughout: this error message is your conclusion from the actual condition being tested. can you log exactly the condition that failed instead?

---

*[storage/reservation_test.go, line 289 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgi_YO0llIfCivCj5A:-KJgi_YO0llIfCivCj5B:-1018986495) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation_test.go#L289)):*
> ```Go
> 		RangeSize: bytesPerReservation,
> 	}).Reserved {
> 		t.Fatalf("could not book a reservation for reservation number 1 (second pass)")
> ```

not new code, but surely "number 1" could be more DRY by using a local for the range ID.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 505 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIe9NJyh0qbNreIcQ8c:-KJgvLFI1q8k8kFmmYlD:1670484775) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L505)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> @mrtracy, why do you think this should be a counter? It seems like a natural gauge to me. (same for `reservedReplicaCount` and the existing `replicaCount`)
</details>
A counter can go up and down, it's not necessarily monotonic; the difference is that counters are adjusted in increments, while a gauge has its entire value overwritten every time.

The advantage of counters is that updates are commutative; if incremental changes are available, counters should be preferred every time.

Gauges are appropriate where the actual value is maintained elsewhere and thus incremental differences are not available; for instance, the "available" capacity is taken from a syscall, so a gauge is appropriate. The "reserved" size, on the other hand, is maintained entirely by cockroach, and a counter should be preferred. 

---

*[storage/store.go, line 1636 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIdOWl_EXFaTIp10k8S:-KJgwFwZ0M5rsb4ZW7Ha:2099925856) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L1636)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I've removed this for now.  I think the best course of action will be to either gossip the metrics themselves (they include a store descriptor already), or to add the metrics to the store descriptor.  @mrtracy and I are reviewing the options.  
> 
> I think some kind of whitelisted subset of the metrics (those that are needed for rebalancing) would be all we need.
</details>
I would really rather remove all metrics from the store descriptor entirely; the store descriptor is a weird mix of semi-permanent configured values (i.e. store ID, attributes, node descriptor) and a dynamic "capacity" field, which in my opinion does not belong on this structure.

I think we should be gossiping a subset of metrics for each store, relevant to rebalancing, on a separate structure.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: all files reviewed at latest revision, 22 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 1636 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIdOWl_EXFaTIp10k8S:-KJhrD6Zu8D0Qt3ZfOr3:1720671756) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L1636)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> I would really rather remove all metrics from the store descriptor entirely; the store descriptor is a weird mix of semi-permanent configured values (i.e. store ID, attributes, node descriptor) and a dynamic "capacity" field, which in my opinion does not belong on this structure.
> 
> I think we should be gossiping a subset of metrics for each store, relevant to rebalancing, on a separate structure.
</details>
It does seem conceptually cleaner and more network efficient to move `capacity` and the new rebalancing metrics to their own new gossiped protobuf, especially if that isn't a large amount of work. At the same time, we really should get the reservation RPC out soon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

BramGruneir committed:
storage: reservation style fixes, rename booking to reservation

BramGruneir commented:
Review status: 0 of 9 files reviewed at latest revision, 22 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 831 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfnxuyD35q9DjA9QtB:-KJgyr8fzamF6HwJ5sTY:-217042822) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/roachpb/api.proto#L831)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It may makes sense to use the larger of the the maximum size and the current size (this may unnecessarily serialize small ranges, but that doesn't seem like a big deal), but either way this will need to be sent as a part of the request - there's not enough information here for the recipient to be able to look up the right zone.
</details>
Yes, I'm working on the request side right now.  

---

*[storage/replica_raftstorage.go, line 715 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfpxQRomSnp7RVQekf:-KJgzvsDMhGd5XZTPgE0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_raftstorage.go#L715)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Yes, that's possible. But as long as `Fill` is a noop when there is no actual reservation it should be fine. It's difficult to tell the difference otherwise.
> 
> Also, I wouldn't use `_ =` when there are no other return values unless you're silencing `errcheck`
</details>
Done.

---

*[storage/replica_test.go, line 5690 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJftN3dENslcWclgNwr:-KJh-9rzKVCzdLSgkCr-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_test.go#L5690)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> +1
</details>
Done.

---

*[storage/replica_test.go, line 5720 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgGHXew5xOCZq1ewl7:-KJh-PBbRmMUufhrLKVy:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/replica_test.go#L5720)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `defer b.Close()`
</details>
Done.

---

*[storage/replica_test.go, line 5698 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJggY4P-enoMtz5h6ti:-KJh-JF0CInbybrTIJBh:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/replica_test.go#L5698)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> just wondering, why not `t.Fatal(err)`? there's already a line number
</details>
Done.

---

*[storage/replica_test.go, line 5722 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJggiDTE_o22kz4oaJw:-KJh-VPSJc8jEpd35mlu:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/replica_test.go#L5722)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---

*[storage/replica_test.go, line 5725 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJggiy7JymQ30JBWL1N:-KJh-VphTaXkboJL7_no:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/replica_test.go#L5725)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---

*[storage/reservation.go, line 96 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgKKdSA7m5CSI00n6M:-KJkRYnCclbvq2uQcR9s:1873084303) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/reservation.go#L96)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this also mutates the argument; is that intended?
</details>
After discussion with @mrtracy, this was not needed and removed.

---

*[storage/reservation.go, line 136 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgKiLLIl1lbRJvmOXe:-KJkYnawOuApRl51bu7-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/reservation.go#L136)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/accomidate/accomodate/
</details>
Done.

---

*[storage/reservation.go, line 87 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJggvjIhSpcBRh05lme:-KJki3aIjdZMLf0rl2QU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation.go#L87)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> doesn't conform to the style guide
</details>
Done.

---

*[storage/reservation.go, line 95 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJghQ2aGZRXy72zum4s:-KJkp8EvedX3h3bdglVf:-460405164) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation.go#L95)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is it a reservation or a booking? this naming is confusing. Also, why pointers in this map?
</details>
because they are also stored in the queue
Also, renamed booking to reservation for clarity.

---

*[storage/reservation.go, line 102 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgh55ueVq0fPQg-44f:-KJkYUy6kb9gPl8g96F4:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation.go#L102)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this comment should explain what a rejected reservation looks like to the caller.
</details>
Done.

---

*[storage/reservation.go, line 107 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJghHC8ZCz-qSiWbUEy:-KJkYW_l7Cy9lvZG7qRN:991488779) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation.go#L107)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this a 
</details>
I've renamed some of the internals to get rid of booking in favour of reservation.

---

*[storage/reservation.go, line 130 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJghcciWw5bAyQgWdFB:-KJkZDsfcyhVsuy7whF3:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation.go#L130)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> did you mean to log the number of live reservations here?
</details>
Done.

---

*[storage/reservation_test.go, line 33 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgi93Up-VcTDXVwO_K:-KJkc8XbD4TAOXHP69oz:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation_test.go#L33)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> does not conform to style guide
</details>
Done.

---

*[storage/reservation_test.go, line 172 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgiI2fLaeViYqirojR:-KJkcLeWLNZLkf2PqVQd:-896019466) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation_test.go#L172)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: this error message is your conclusion from the actual condition being tested. can you log exactly the condition that failed instead?
</details>
The fail case is logged, at v2 level.  But I tried to make the error case message cleaner.

---

*[storage/reservation_test.go, line 289 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgi_YO0llIfCivCj5A:-KJkpQradIhTtt10NL9E:-2014200741) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation_test.go#L289)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not new code, but surely "number 1" could be more DRY by using a local for the range ID.
</details>
I like the explicitness of it as is.

---

*[storage/store.go, line 76 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJfsOjq0egTFgN5qnbS:-KJkhfsHxrdsr_Cd92TU:2049923341) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb01fda3406cff171e93ae781254f95c0512a63f/storage/store.go#L76)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Or `1<<30`, which appears to be the more common style in our code. (s/gigibyte/GiB/)
> 
> I would make these numbers really small to start to prioritize stability (say 5 and 250MB); we can consider increasing them if rebalancing is still too slow (or when we have streaming snapshots).
</details>
Done.  I plan to make these values environment variables for faster testing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
All changes based on comments applied.  PTAL

tamird commented:
Reviewed 6 of 7 files at r3, 3 of 3 files at r4, 1 of 5 files at r6, 2 of 7 files at r8, 1 of 4 files at r9, 1 of 3 files at r10, 2 of 5 files at r11, 3 of 3 files at r12.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[storage/reservation_test.go, line 172 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgiI2fLaeViYqirojR:-KJktkf0os3pP7qJGWGn:1644450453) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation_test.go#L172)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> The fail case is logged, at v2 level.  But I tried to make the error case message cleaner.
</details>
I don't get it. Why not "expected reservation to fail due to lack of capacity, but it succeeded".

---

*[storage/reservation_test.go, line 110 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJkteVNFxy7nuFBUACR:-KJkteVNFxy7nuFBUACS:284059936) ([raw file](https://github.com/cockroachdb/cockroach/blob/3301883a08b302ddd9b0d5c0985e40dcdc5d9b84/storage/reservation_test.go#L110)):*
> ```Go
> 			}
> 		} else {
> 			// Fill the resrvation.
> ```

typo throughout

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 7 of 9 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[storage/reservation_test.go, line 172 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJgiI2fLaeViYqirojR:-KJl8gBXhNF5fNhZRKDH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c068f563acfd4b7eb742ebeffc6e108f5960b88d/storage/reservation_test.go#L172)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I don't get it. Why not "expected reservation to fail due to lack of capacity, but it succeeded".
</details>
Done.

---

*[storage/reservation_test.go, line 110 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJkteVNFxy7nuFBUACR:-KJl8CwQGH13xT1GGUd8:1791321975) ([raw file](https://github.com/cockroachdb/cockroach/blob/3301883a08b302ddd9b0d5c0985e40dcdc5d9b84/storage/reservation_test.go#L110)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> typo throughout
</details>
Done.  Sorry about that.  My spell checker was still not working properly and is starting to really piss me off.  I think I've fixed it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r13.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[storage/reservation_test.go, line 173 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJlEMZ1qUbsJ2gNYHUG:-KJlEMZ2IeZB-WZVgJzA:-1849673022) ([raw file](https://github.com/cockroachdb/cockroach/blob/56563fc247e00b37b8588f6e145c5f358423c840/storage/reservation_test.go#L173)):*
> ```Go
> 
> 	if b.Reserve(overfilledReq).Reserved {
> 		t.Errorf("expected reservation to fail due to too lack of available disk space, but it succeeded")
> ```

throughout: "to too lack"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 8 of 9 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[storage/reservation_test.go, line 173 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KJlEMZ1qUbsJ2gNYHUG:-KJlLLizH9zuhAMIfmJB:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/56563fc247e00b37b8588f6e145c5f358423c840/storage/reservation_test.go#L173)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: "to too lack"
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r14, 2 of 7 files at r16, 1 of 3 files at r17, 1 of 3 files at r18, 2 of 5 files at r19, 3 of 3 files at r20.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> All changes based on comments applied.  PTAL
</details>


---

Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-:-KJlZma1teVHvfOiC-A5:1417088721)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 505 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904#-KIe9NJyh0qbNreIcQ8c:-KK0eVGDPRwUm5L-p3Up:1399981613) ([raw file](https://github.com/cockroachdb/cockroach/blob/d65cc9421208dfac85994cf7e74b9b77ef9c5af5/storage/store.go#L505)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> A counter can go up and down, it's not necessarily monotonic; the difference is that counters are adjusted in increments, while a gauge has its entire value overwritten every time.
> 
> The advantage of counters is that updates are commutative; if incremental changes are available, counters should be preferred every time.
> 
> Gauges are appropriate where the actual value is maintained elsewhere and thus incremental differences are not available; for instance, the "available" capacity is taken from a syscall, so a gauge is appropriate. The "reserved" size, on the other hand, is maintained entirely by cockroach, and a counter should be preferred. 
</details>
In my mind the metric types are also signals to the monitoring framework that indicate how these metrics are to be used. The key difference between a counter and a gauge to me is that counters are used to compute their derivative, while gauges are displayed directly (or in simple aggregates like sum, max, or mean). So that's why I favor gauges here even when we're updating them by incrementing (we could add Increment methods to our Gauge type).

It's difficult to take advantage of the fact that most counter updates are commutative because of the reset to zero that occurs when a process restarts; this is not commutative with inc/decrements.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6904)*
<!-- Sent from Reviewable.io -->
